src/python/pants/jvm/compile.py

```diff
@@ -242,7 +242,7 @@ def __init__(
         dependencies: Iterable[ClasspathEntry] = (),
     ):
         self.digest = digest
-        self.filenames = tuple(filenames)
+        self.filenames = tuple(sorted(filenames))
```

Comment:
As mentioned in the docstring, this needs to preserve the input ordering (to prevent adjusting the output order from e.g. Coursier). So (relevant) callsites will probably need to updated instead.